### PR TITLE
Simple Example for DifferentialEquations integration

### PR DESCRIPTION
**Pull request by termi-official**
Turns out implementing a consistent multi-level Newton-Raphson for DifferentialEquations is quite tricky ( #325  is blocked for now ). Since several users asked about examples for the integration, here a simpler example.

## Open TODOs

* [x] Implement incompressibles Stokes flow
* [x] ~~Fix DAE initialization~~ Correct initialization of the algebraic portion is nontrivia, since the semi-discretization has Hessenberg index 2. DiffEq just has methods for index 1 dae initialization, hence we just start with a trivial (zero) initial condition.
* [x] Implement incompressible Navier-Stokes flow
* [x] ~~Simple mesh loader to change regular grid with a domain with hole~~ This can be done as a follow up PR. Currently the mesh is generated manually.
* [x] Fix CI pipeline (update somehow broke plotting)
* [ ] ~~new apply_zero! dispatch to zero out just a sub-field~~ -> future work... 

## Non-addressed issues
* Mesh IO: the example currently generates the mesh manually, which is suboptimal. Improving this situation is left for follow up  PR and out of scope for this PR.
* Solver: A direct sparse solver has been utilized without bandwith optimizing reordering of the matrix, leading to an immense amount of memory necessary. Also the Stokes operator is just semi-definite, which may lead to additional issues with the direct solver. A preconditioned GMRes should be deployed through a future PR.

### CODE REVIEW

**Comment by termi-official** [comment]
Vortex shedding works. :) Cleanup and documentation will follow.
vortex-shedding

**Comment by KristofferC** [comment]
Epic!

**Comment by termi-official** [comment]
Almost finished. This example basically shows how timestep with DifferentialEquations.jl with a semi-discretization of problems which can be stated in mass matrix form and time-dependent Dirichlet boundary conditions.

I think I may need one or two more iterations on the explanation. A major weakness for now is that I have to overwrite one of the internal methods from the OrdinaryDiffEq package (OrdinaryDiffEq.initialize!) to apply the Dirichlet BC correctly in the time-dependent case. If the Dirichlet BC does not depend on time, everything should also work fine without hooking in this method. Also, I can not guarantee that my patch of OrdinaryDiffEq.initialize! works for all timesteppers as it may mess with assumptions taken in the corresponding OrdinaryDiffEq.perform_step! implementations.

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:11]:
```suggestion
# In this example we focus on a simple but visually appealing problem from
# fluid dynamics, namely vortex shedding. This problem is also known as
# [von-Karman vortex streets](https://en.wikipedia.org/wiki/K%C3%A1rm%C3%A1n_vortex_street). Within this example, we show how to utilize [DifferentialEquations.jl]()
# in tandem with [Ferrite.jl](). To keep things simple we use a general approach
# to discretize the system: [Chorin's method](https://en.wikipedia.org/wiki/Projection_method_(fluid_dynamics)#Chorin's_projection_method). A performant projection method for the Navier-Stokes equations.
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:13]: maybe add in this section a dummy example how mass matrix form looks like

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:37]:
```suggestion
# operator, so for any solution $p$ the function $p + c$ is also an admissible solution, if
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:46]:
```suggestion
# channel have no-slip conditions, i.e. $v = (0,0)$, while the right boundary has the do-nothing boundary condtion
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:75]:
```suggestion
# Since we do not need the complete DifferentialEquations suite, we just load the required ODE infrastructure, which can also handle
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:167]: maybe add here once again the weak form and mark with `underbrace` which parts correspond to stokes, advective and mass terms

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:210]: this part isn't displayed properly in the docs deployment, maybe try `bmatrix`?

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:247]:
```suggestion
            # Pressure + Incompressibility term $B$ - note the symmetry.
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:238]:
```suggestion
            # Viscosity term $A$
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:326]:
```suggestion
    A::Union{Nothing, SparseMatrixCSC}
```
I guess, can also be parametrized, at least some information should be beneficial for the compiler, see https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:328]: one white line between each function declaration would look better imho. Does the function need to have this `p::` closure? It looks not really beginner friendly, but I guess we are constrained to this since DiffEq calls this?

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:308]:
```suggestion
    # The update of `u` takes `uprev + z` or `tmp + z` most of the time, so we have
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:345]:
```suggestion
# feasible for semi-discretization of finite element models. We communicate
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:372]:
```suggestion
                # Note that the other the gradient term is now on the left, which is the correct thing to do here.
```
not sure, but I guess `other`

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:375]:
```suggestion
                # [(v \cdot \nabla) v] = v_i \partial_i v_j = \partial_i v_i v_j = (\nabla v) v
```

Otherwise you'd have to write 
```md
               # [(v \cdot \nabla) v]_{ij} = v_i \partial_i v_j = \partial_i v_i v_j = (\nabla v) v
```

Because in tensor notation you only apply indices to a bracket if you deconstruct every inner running index, in order to show the order of the tensor. At least this is what is done in the mechanics community and "canonical" in the sense of tensor notation. Sometimes, also utilized to denote a derivative with running index to a bracket by e.g. `(v_i v_j)_{,k}` would be the derivative with running index `k` to the inner arguments. Nonetheless, the indices should start first with i and secondly with j, by that you get naturally the correct nabla formulation, you can also be very explicit by 

```md
               # [(v \cdot \nabla) v] = v_i \boldsymbol{e}_i \cdot \partial_j \boldsymbol{e}_j v_k \boldsymbol{e}_k = v_i \partial_j \delta_{ij} v_k \boldsymbol{e}_k = v_i \partial_i \boldsymbol{e}_k v_k \boldsymbol{e}_k = \partial_i v_i v_j = (\nabla v) v
```

**Review by koehlerson — COMMENTED** [review]
added a few corrections and suggestions :)

**Comment by fredrikekre** [comment]
> A major weakness for now is that I have to overwrite one of the internal methods from the OrdinaryDiffEq package (OrdinaryDiffEq.initialize!) to apply the Dirichlet BC correctly in the time-dependent case.

Would it make sense to make a feature request about this then?

**Comment by termi-official** [comment]
Thanks for the feedback @koehlerson , I will address your points later.

@fredrikekre regarding the feature request I am not sure what exactly I should request here. Technically it is at the current state already possible to not overwrite the initialize! method. I found here 3 ways
A) useing Cassette.jl and patch the initialize! method (for Newton) instead
B) developing a wrapper for DiffEq's Newton
C) utilize operators for the BCs

I think A is a bit too hacky, so I didn't wanted to go this route.

I tried B but failed for now, because the NL solver appears at many points in the code. Should not be impossible tho, but will take more time. A problem with this approach is, that not every solver utilizes the this Newton loop. The NL solve framework is quite fragmented atm, as I discussed with the maintainer (Chris) on Slack, and there is a chance that this will be refactored in the future. For the time steppers which do not utilize the internal NL solve framework (e.g. explicit and Rosenbrock-type methods) I could not find a way to enforce time-dependent DBCs at all. I also cannot guarantee in the current implementation, that all time steppers which utilize the Newton will work, because I rely on the assumption that the time stepper updates the solution $u$ either via $uprev + dt_{stage}*z$ or $tmp + dt_{stage}*z$.

Regarding C: For linear problems we could adopt the trick from [DiffEqOperators.jl](https://github.com/SciML/DiffEqOperators.jl), but I also could not find a way to generalize this to nonlinear problems.


To elaborate a bit more on why I pursuit B over the alternatives. The goal of #325 is to develop a staggered/multilevel-type Newton solver integration. These algorithms are efficient when we have nonlinear equations at the integration points, as e.g. in several plasticity models. With this PR I wanted to explore the simpler case, when we do not have any additional equations at the integration points.

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:328]: For the second call we need this closure, because the function is stateful, carrying the sparse LU decomposition and the constraint handler.

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:326]: Also, for some reason DiffEq chokes if we do this.

Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:326]: hm but then it should still be possible to constraint the type by some parametrization e.g. `Union{Nothing,T}` and T is then `<: Union{MatrixFree,AbstractMatrix}`. Having no constraint doesn't feel good to me, since beginner will take the code and most likely copy paste it

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:328]: Sorry, question was misleading, so we define here something like a functor right? I'm just wondering if we can instead use a normal function, otherwise we should at least explain why this is needed and how it works with a link to https://docs.julialang.org/en/v1/manual/methods/index.html#Function-like-objects-1

Beginners will have a hard time figuring out what is happening in the 
```julia
function (p::FerriteLinSolve)()
```
part

Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:328]: I tried to put in a normal function for the first one and got a very strange type error. Not sure what exactly is breaking. If you want you can take a closer look on this one, I will push my remaining fixes in a few minutes.

For this I just derived some quick and dirty example from https://diffeq.sciml.ai/stable/features/linear_nonlinear/#Implementing-Your-Own-LinSolve:-How-LinSolveFactorize-Was-Created

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:326]: No constraints should be fine here. I do not think that possible type instabilities here could lead to bottlenecks - usually either the evaluation of the right hand side, assembly or the solve is the bottleneck.

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:209]: Is this also resolved?

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:370]: It's primarly about providing good pratical code examples and having unconstrained types in structs is not advised according to the docs. I benchmarked it, its just 4s difference 56s vs 52s with parametrization. But still, 7% by a trivial change. 

In case you want to change it:
```julia
FerriteLinSolve(ch) = FerriteLinSolve(ch,lu,lu(sparse(ones(1,1))))

mutable struct FerriteLinSolve{CH,F,T<:Factorization}
    ch::CH
    factorization::F
    A::T
end
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:209]: yes :)

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:370]: Okay, I agree on this point, thanks!

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl]: We can just leave it as it is. But I would recommend to mention the link to the docs or at least mention functor somewhere in between

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:17]:
```suggestion
# discretization technique - in our case FEM. The mass matrix form of ODEs and DAEs
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:75]:
```suggestion
#                                  0 &= \left(\int \nabla \cdot v \right) \psi
```
not necessary but it's r/mildlyinfuriating without `\left\right`:D

Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:75]: oh wait, is the opening bracket actually intended before the integral? Or should it be in front of the nabla?

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:364]:
```suggestion
p = RHSparams(K, ch, dh, cellvalues_v)

```
white line, otherwise function declaration and RHSparams stuff is in one Documenter code block

Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:364]: grafik

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:435]:
```suggestion

```
legacy

File: docs/src/literate/ns_vs_diffeq.jl
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:75]: The integral should be on the outside

**Comment by koehlerson** [comment]
LGTM :)

**Comment by termi-official** [comment]
Thanks for the detailed review Maxi! This one should be ready - however some future work is left open.

1. Mesh IO / refactoring of the mesh generator: Currently I just hack together a holed domain. Probably not the best way. Should definitely be addressed once we have some mesh IO. Or the mesh generator should be refactored.
2. Preconditioner+iterative solver: Once this is merged I will look into the dof handler for a follow up PR to address the assembly mode. Currently we assemble blocked systems as `ABABAB....`, where `A` and `B` are representing the fields. In the case of incompressible Navier-Stokes A=velocity and B=pressure. However, some preconditioners require the system to be in the form `AAAA.....BBBB....`, see e.g. https://www.dealii.org/current/doxygen/deal.II/step_57.html#TheSolverandPreconditioner

**Comment by KristofferC** [comment]
> he example has tests included, but for now I didn't wanted to introduce the fat DifferentialEquations dependency into Ferrite just for the test (currently the dep is just in the doc).

You can add test specific dependencies: https://github.com/Ferrite-FEM/Ferrite.jl/blob/b44132c6d9a2e4b0dbadc582635a6bb3dff086b6/Project.toml#L18-L33

**Comment by fredrikekre** [comment]
Good to go? Nice that you managed without overwriting DiffEq internals!

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:453]: renders in docs, should be hidden

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:115]:
```suggestion
# First we load Ferrite and some other packages we need
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:159]:
```suggestion
# We have to utilize the same quadrature rule for the velocity as for the pressure, because in the weak form the
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:340]:
```suggestion
# and finally set the r.h.s. to zero on every constraint. This way we obtain a
```

File: docs/src/literate/ns_vs_diffeq.jl
Comment (koehlerson) [inline docs/src/literate/ns_vs_diffeq.jl:339]:
```suggestion
# discretized right hand side (r.h.s.) of the Navier-Stokes equations with this vector
```

**Review by koehlerson — COMMENTED** [review]
last few bits before merging

**Comment by termi-official** [comment]
Should be fine @fredrikekre . Docs builds properly on my machine. Dunno why the the last push did not trigger the CI. Once this is merged I will open up the follow up issues.

**Comment by fredrikekre** [comment]
Seems to be a github issue: https://www.githubstatus.com/

**Review by koehlerson — APPROVED** [review]

**Comment by fredrikekre** [comment]
"Simple Example" 😛

**Comment by KristofferC** [comment]
7drHiqr

**Pull request merged**
